src/data-model-providers/codegen/ServerClusterInterfaceRegistry.h

```diff
@@ -0,0 +1,140 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/ConcreteClusterPath.h>
+#include <app/server-cluster/ServerClusterInterface.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/DataModelTypes.h>
+
+namespace chip {
+namespace app {
+
+/// Allows registering and retrieving ServerClusterInterface instances for specific cluster paths.
+class ServerClusterInterfaceRegistry
+{
+private:
+    struct RegisteredServerClusterInterface;
+
+public:
+    /// represents an iterable list of clusters
+    class ClustersList
+    {
+    public:
+        class Iterator
+        {
+        public:
+            Iterator(RegisteredServerClusterInterface * interface) : mInterface(interface) {}
+
+            Iterator & operator++()
+            {
+                if (mInterface != nullptr)
+                {
+                    mInterface = mInterface->next;
+                }
+                return *this;
+            }
+            bool operator==(const Iterator & other) const { return mInterface == other.mInterface; }
+            bool operator!=(const Iterator & other) const { return mInterface != other.mInterface; }
+            ServerClusterInterface * operator*() { return mInterface->serverClusterInterface; }
+
+        private:
+            RegisteredServerClusterInterface * mInterface;
+        };
+
+        ClustersList(RegisteredServerClusterInterface * start) : mStart(start) {}
+        Iterator begin() { return mStart; }
+        Iterator end() { return nullptr; }
+
+    private:
+        RegisteredServerClusterInterface * mStart;
+    };
+
+    ~ServerClusterInterfaceRegistry();
+
+    /// Associate a specific interface with the given endpoint.
+    ///
+    /// There can be only a single registration for a given `endpointId/clusterId` path.
+    /// A registration will return an error if a registration already exists on
+    /// the given `endpointId/cluster->GetClusterID()` already exists
+    ///
+    /// Registrations need a valid `endpointId` and `cluster->GetClusterID()` MUST be a valid cluster id.
+    [[nodiscard]] CHIP_ERROR Register(EndpointId endpointId, ServerClusterInterface * cluster);
+
+    /// Remove an existing registration for a given endpoint/cluster path.
+    ///
+    /// Returns the previous registration if any exists (or nullptr if nothing
+    /// to unregister)
+    ServerClusterInterface * Unregister(const ConcreteClusterPath & path);
+
+    /// Return the interface registered for the given cluster path or nullptr if one does not exist
+    ServerClusterInterface * Get(const ConcreteClusterPath & path);
+
+    /// ClustersList is valid as a snapshot only and should not be saved.
```

Comment:
It's not a snapshot in any way: it's only valid until someone modifies the registry, right?

Reply:
It is assumed to become invalid whenever the interface is changed in some way (specifically if someone unregisters the element the list points at, things may break badly). I can rephrase.